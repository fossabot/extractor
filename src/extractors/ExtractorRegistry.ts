import Registry from "../common/Registry";
import BowerJsonExtractor from "./BowerJsonExtractor";
import BuildGradleExtractor from "./BuildGradleExtractor";
import CargoTomlExtractor from "./CargoTomlExtractor";
import ComposerJsonExtractor from "./ComposerJsonExtractor";
import Extractor from "./Extractor";
import GodepsJsonExtractor from "./GodepsJsonExtractor";
import GoModExtractor from "./GoModExtractor";
import GopkgTomlExtractor from "./GopkgTomlExtractor";
import IvyXmlExtractor from "./IvyXmlExtractor";
import PackageJsonExtractor from "./PackageJsonExtractor";
import PomXmlExtractor from "./PomXmlExtractor";
import VendorConfExtractor from "./VendorConfExtractor";

const ExtractorRegistry = new Registry<Extractor>("Extractor");

ExtractorRegistry.registerAll({
    BuildGradleExtractor: async (_) => new BuildGradleExtractor(),
    CargoTomlExtractor: async (_) => new CargoTomlExtractor(),
    ComposerJsonExtractor: async (_) => new ComposerJsonExtractor(),
    GodepsJsonExtractor: async (_) => new GodepsJsonExtractor(),
    GoModExtractor: async (_) => new GoModExtractor(),
    GopkgTomlExtractor: async (_) => new GopkgTomlExtractor(),
    IvyXmlExtractor: async (_) => new IvyXmlExtractor(),
    PackageJsonExtractor: async (_) => new PackageJsonExtractor(),
    PomXmlExtractor: async (_) => new PomXmlExtractor(),
    BowerJsonExtractor: async (_) => new BowerJsonExtractor(),
    VendorConfExtractor: async (_) => new VendorConfExtractor(),
});

export default ExtractorRegistry;
