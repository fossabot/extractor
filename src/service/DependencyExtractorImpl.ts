import {ServerUnaryCall} from "grpc";
import {DependencyManagementFile} from "../../api/deps";
import {ExtractRequest, ExtractResponse, MatchRequest, MatchResponse} from "../../api/extractor";
import Extractor from "../extractors/Extractor";
import ExtractorFile from "../extractors/ExtractorFile";
import AsyncDependencyExtractor from "./AsyncDependencyExtractor";

function constructTree(separator: string, paths: string[]): any {
    const root: any = {};

    paths.forEach((key) => {
        const parts = key.split(separator);

        let ptr = root;
        let i = 0;
        for (; i < parts.length - 1; i++) {
            const part = parts[i];
            if (!ptr[part]) {
                ptr[part] = {};
            }
            ptr = ptr[part];
        }
        ptr[parts[i]] = key;
    });

    return root;
}

export default class DependencyExtractorImpl implements AsyncDependencyExtractor {
    private readonly extractors: Extractor[];

    constructor(extractors: Extractor[]) {
        this.extractors = extractors;
    }

    public matchInternal(separator: string, paths: string[]): string[] {
        const root = constructTree(separator, paths);

        let level = [ root ];
        const matchedPaths = [];

        while (level.length > 0) {
            const size = level.length;

            for (let i = 0; i < size; i++) {
                const dir = level.shift();

                this.extractors
                    .filter((extractor) =>
                        extractor.requires()
                            .map((req) => dir[req] && typeof dir[req] === "string")
                            .reduce((last, current) => last && current))
                    .forEach((extractor) =>
                        extractor.requires()
                            .forEach((req) => matchedPaths.push(dir[req])));

                const nextLevel = Object.keys(dir)
                    .map((name) => dir[name])
                    .filter((val) => typeof val !== "string");

                level = level.concat(nextLevel);
            }
        }

        return matchedPaths;
    }

    public async match(call: ServerUnaryCall<MatchRequest>): Promise<MatchResponse> {
        const { separator, paths } = call.request;

        return {
            matchedPaths: this.matchInternal(separator, paths),
        };
    }

    public async extractInternal(
        url: string,
        separator: string,
        fileContents: { [key: string]: string },
    ): Promise<DependencyManagementFile[]> {
        const matchedPaths = this.matchInternal(separator, Object.keys(fileContents));

        const root = constructTree(separator, matchedPaths);

        let level = [ root ];
        let managementFilePromises: Array<Promise<DependencyManagementFile>> = [];

        while (level.length > 0) {
            const size = level.length;

            for (let i = 0; i < size; i++) {
                const dir = level.shift();

                const nextManagementFilePromises = this.extractors
                    .filter((extractor) =>
                        extractor.requires()
                            .map((req) => dir[req] && typeof dir[req] === "string")
                            .reduce((last, current) => last && current))
                    .map((extractor) => {
                        const files = {};
                        extractor.requires().forEach((req) => {
                            const key = dir[req];
                            files[req] = new ExtractorFile(fileContents[key]);
                        });
                        return extractor.extract(url, files);
                    });

                managementFilePromises = managementFilePromises.concat(nextManagementFilePromises);

                const nextLevel = Object.keys(dir)
                    .map((name) => dir[name])
                    .filter((val) => typeof val !== "string");

                level = level.concat(nextLevel);
            }
        }

        let managementFiles = await Promise.all(managementFilePromises);
        managementFiles = managementFiles.filter((f) => !!f);

        return managementFiles;
    }

    public async extract(call: ServerUnaryCall<ExtractRequest>): Promise<ExtractResponse> {
        const { url, separator, fileContents } = call.request;

        const managementFiles = await this.extractInternal(url, separator, fileContents);

        return {
            managementFiles,
        };
    }
}
