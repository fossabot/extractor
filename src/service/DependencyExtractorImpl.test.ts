import fs = require("fs");
import path = require("path");
import ExtractorRegistry from "../extractors/ExtractorRegistry";
import DependencyExtractorImpl from "./DependencyExtractorImpl";

const fsp = fs.promises;

const files = [
    "BuildGradleExtractor.test.ts",
    "BuildGradleExtractor.ts",
    "CargoTomlExtractor.test.ts",
    "CargoTomlExtractor.ts",
    "ComposerJsonExtractor.test.ts",
    "ComposerJsonExtractor.ts",
    "Extractor.ts",
    "ExtractorFile.ts",
    "ExtractorRegistry.ts",
    "GoModExtractor.test.ts",
    "GoModExtractor.ts",
    "GodepsJsonExtractor.test.ts",
    "GodepsJsonExtractor.ts",
    "GopkgTomlExtractor.test.ts",
    "GopkgTomlExtractor.ts",
    "IvyXmlExtractor.test.ts",
    "IvyXmlExtractor.ts",
    "PackageJsonExtractor.test.ts",
    "PackageJsonExtractor.ts",
    "PomXmlExtractor.test.ts",
    "PomXmlExtractor.ts",
    "testdata/Cargo.toml",
    "testdata/Godeps.json",
    "testdata/Gopkg.toml",
    "testdata/build.gradle",
    "testdata/composer.json",
    "testdata/go.mod",
    "testdata/go.sum",
    "testdata/ivy.xml",
    "testdata/package.json",
    "testdata/pom.xml",
    "testdata/settings.gradle",
];

describe("DependencyExtractorImpl", () => {
    test("fullParse", async () => {
        const extractorPromises = ExtractorRegistry.known()
            .map((registry) => ExtractorRegistry.resolve(registry, null));

        const extractors = await Promise.all(extractorPromises);

        const extractorImpl = new DependencyExtractorImpl(extractors);

        const matched = extractorImpl.matchInternal(path.sep, files);

        expect(matched).toMatchSnapshot();

        const extractorsDir = path.resolve(__dirname, "../extractors");

        const fileContents = {};
        const promises = matched.map((match) =>
            fsp.readFile(path.join(extractorsDir, match))
                .then((buf) => buf.toString())
                .then((content) => fileContents[match] = content));

        await Promise.all(promises);

        const dependencyManagementFiles = await extractorImpl.extractInternal(
            "git@github.com:deps-cloud/extractor.git", path.sep, fileContents);

        expect(dependencyManagementFiles).toMatchSnapshot();
    });
});
