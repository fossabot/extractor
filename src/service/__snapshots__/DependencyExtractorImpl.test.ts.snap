// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependencyExtractorImpl fullParse 1`] = `
Array [
  "testdata/build.gradle",
  "testdata/settings.gradle",
  "testdata/Cargo.toml",
  "testdata/composer.json",
  "testdata/Godeps.json",
  "testdata/go.mod",
  "testdata/Gopkg.toml",
  "testdata/ivy.xml",
  "testdata/package.json",
  "testdata/pom.xml",
]
`;

exports[`DependencyExtractorImpl fullParse 2`] = `
Array [
  Object {
    "dependencies": Array [
      Object {
        "module": "parent_module",
        "organization": "parent_organization",
        "scopes": Array [
          "parent",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "kubernetes-client",
        "organization": "io.fabric8",
        "scopes": Array [
          "compile",
          "annotationProcessor",
        ],
        "versionConstraint": "4.1.3",
      },
      Object {
        "module": "commons-collections4",
        "organization": "org.apache.commons",
        "scopes": Array [
          "compile",
        ],
        "versionConstraint": "4.2",
      },
      Object {
        "module": "commons-text",
        "organization": "org.apache.commons",
        "scopes": Array [
          "compile",
        ],
        "versionConstraint": "1.6",
      },
      Object {
        "module": "jackson-core",
        "organization": "com.fasterxml.jackson.core",
        "scopes": Array [
          "compile",
        ],
        "versionConstraint": "2.9.7",
      },
      Object {
        "module": "spring-boot-starter",
        "organization": "org.springframework.boot",
        "scopes": Array [
          "compile",
        ],
        "versionConstraint": "2.1.1.RELEASE",
      },
      Object {
        "module": "spring-boot-starter-log4j2",
        "organization": "org.springframework.boot",
        "scopes": Array [
          "compile",
        ],
        "versionConstraint": "2.1.1.RELEASE",
      },
      Object {
        "module": "builder-annotations",
        "organization": "io.sundr",
        "scopes": Array [
          "compileOnly",
        ],
        "versionConstraint": "0.14.7",
      },
      Object {
        "module": "mockito-core",
        "organization": "org.mockito",
        "scopes": Array [
          "testCompile",
        ],
        "versionConstraint": "2.24.0",
      },
      Object {
        "module": "junit-jupiter-api",
        "organization": "org.junit.jupiter",
        "scopes": Array [
          "testCompile",
        ],
        "versionConstraint": "5.4.0",
      },
      Object {
        "module": "junit-jupiter-engine",
        "organization": "org.junit.jupiter",
        "scopes": Array [
          "testRuntime",
        ],
        "versionConstraint": "5.4.0",
      },
    ],
    "language": "java",
    "module": "module",
    "organization": "organization",
    "system": "gradle",
    "version": "1.2",
  },
  Object {
    "dependencies": Array [
      Object {
        "module": "bytes",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "clap",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "futures",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "hyper",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "log",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "ordermap",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "pretty_env_logger",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "rand",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "rustls",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": "make_server_sni_public",
      },
      Object {
        "module": "serde",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "serde_derive",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "serde_json",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "serde_yaml",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "tacho",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "tokio-core",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "tokio-io",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "tokio-service",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "tokio-timer",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "url",
        "organization": "_",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": undefined,
      },
    ],
    "language": "rust",
    "module": "linkerd-tcp",
    "organization": "_",
    "system": "cargo",
    "version": "0.1.1",
  },
  Object {
    "dependencies": Array [
      Object {
        "module": "repoModule",
        "organization": "repoOrganization",
        "versionConstraint": "repoVersion",
      },
      Object {
        "module": "requireModule",
        "organization": "requireOrganization",
        "versionConstraint": "requireVersionConstraint",
      },
      Object {
        "module": "requireDevModule",
        "organization": "requireDevOrganization",
        "versionConstraint": "requireDevVersionConstraint",
      },
    ],
    "language": "php",
    "module": "module",
    "organization": "organization",
    "system": "composer",
    "version": "version",
  },
  Object {
    "dependencies": Array [
      Object {
        "module": "kr/fs",
        "organization": "github.com",
        "versionConstraint": "2788f0dbd16903de03cb8186e5c7d97b69ad387b",
      },
      Object {
        "module": "kr/pretty",
        "organization": "github.com",
        "versionConstraint": "f31442d60e51465c69811e2107ae978868dbea5c",
      },
      Object {
        "module": "kr/text",
        "organization": "github.com",
        "versionConstraint": "6807e777504f54ad073ecef66747de158294b639",
      },
      Object {
        "module": "pmezard/go-difflib/difflib",
        "organization": "github.com",
        "versionConstraint": "f78a839676152fd9f4863704f5d516195c18fc14",
      },
      Object {
        "module": "x/tools/go/vcs",
        "organization": "golang.org",
        "versionConstraint": "1f1b3322f67af76803c942fd237291538ec68262",
      },
    ],
    "language": "go",
    "module": "tools/godep",
    "organization": "github.com",
    "system": "godeps",
    "version": "",
  },
  Object {
    "dependencies": Array [
      Object {
        "module": "gogo/protobuf",
        "organization": "github.com",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": "v1.2.1",
      },
      Object {
        "module": "deps-cloud/rds",
        "organization": "github.com",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": "v0.0.2",
      },
      Object {
        "module": "sirupsen/logrus",
        "organization": "github.com",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": "v1.3.0",
      },
      Object {
        "module": "x/net",
        "organization": "golang.org",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": "v0.0.0-20190213061140-3a22650c66bd",
      },
      Object {
        "module": "grpc",
        "organization": "google.golang.org",
        "scopes": Array [
          "direct",
        ],
        "versionConstraint": "v1.18.0",
      },
    ],
    "language": "go",
    "module": "deps-cloud/finch",
    "organization": "github.com",
    "system": "vgo",
    "version": "latest",
  },
  Object {
    "dependencies": Array [
      Object {
        "module": "go",
        "organization": "cloud.google.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": ">=0.20.0, <1.0.0",
      },
      Object {
        "module": "go.opencensus.io",
        "organization": "",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": ">=0.15.0",
      },
      Object {
        "module": "exporter/stackdriver",
        "organization": "contrib.go.opencensus.io",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": ">=0.2.0",
      },
      Object {
        "module": "copilot",
        "organization": "code.cloudfoundry.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "DataDog/datadog-go",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "2.0.0",
      },
      Object {
        "module": "alicebob/miniredis",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "2.3.2",
      },
      Object {
        "module": "aws/aws-sdk-go",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.13.24",
      },
      Object {
        "module": "cactus/go-statsd-client",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "3.1.1",
      },
      Object {
        "module": "cenkalti/backoff",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "2.0.0",
      },
      Object {
        "module": "circonus-labs/circonus-gometrics",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "2.1.0",
      },
      Object {
        "module": "coreos/go-oidc",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "v2",
      },
      Object {
        "module": "davecgh/go-spew",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.1.0",
      },
      Object {
        "module": "envoyproxy/go-control-plane",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "^0.6.7",
      },
      Object {
        "module": "fluent/fluent-logger-golang",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.3.0",
      },
      Object {
        "module": "ghodss/yaml",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.0.0",
      },
      Object {
        "module": "go-redis/redis",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "6.10.2",
      },
      Object {
        "module": "gogo/googleapis",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "gogo/protobuf",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "^1.1.1",
      },
      Object {
        "module": "golang/protobuf",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.1.0",
      },
      Object {
        "module": "golang/sync",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "google/go-github",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "15.0.0",
      },
      Object {
        "module": "google/uuid",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "0.2.0",
      },
      Object {
        "module": "googleapis/gax-go",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "2.0.0",
      },
      Object {
        "module": "gorilla/mux",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.6.1",
      },
      Object {
        "module": "gorilla/websocket",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.2.0",
      },
      Object {
        "module": "grpc-ecosystem/go-grpc-middleware",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "grpc-ecosystem/go-grpc-prometheus",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.1.0",
      },
      Object {
        "module": "grpc-ecosystem/grpc-opentracing",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "hashicorp/consul",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.3.0",
      },
      Object {
        "module": "hashicorp/go-multierror",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "hashicorp/vault",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "0.10.0",
      },
      Object {
        "module": "howeyc/fsnotify",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "0.9.0",
      },
      Object {
        "module": "natefinch/lumberjack",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "2.1.0",
      },
      Object {
        "module": "onsi/gomega",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.3.0",
      },
      Object {
        "module": "open-policy-agent/opa",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "0.8.2",
      },
      Object {
        "module": "openshift/api",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "opentracing/opentracing-go",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.0.2",
      },
      Object {
        "module": "pmezard/go-difflib",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.0.0",
      },
      Object {
        "module": "prometheus/client_golang",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "0.9.2",
      },
      Object {
        "module": "prometheus/client_model",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "prometheus/common",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "prometheus/prom2json",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "satori/go.uuid",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.2.0",
      },
      Object {
        "module": "spf13/cobra",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "0.0.2",
      },
      Object {
        "module": "stretchr/testify",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.2.1",
      },
      Object {
        "module": "uber/jaeger-client-go",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "atomic",
        "organization": "go.uber.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.2.0",
      },
      Object {
        "module": "multierr",
        "organization": "go.uber.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.1.0",
      },
      Object {
        "module": "zap",
        "organization": "go.uber.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "1.7.1",
      },
      Object {
        "module": "x/net",
        "organization": "golang.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "x/oauth2",
        "organization": "golang.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "x/time",
        "organization": "golang.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "x/tools",
        "organization": "golang.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "api",
        "organization": "google.golang.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "genproto",
        "organization": "google.golang.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "grpc",
        "organization": "google.golang.org",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "^1.13.0",
      },
      Object {
        "module": "yaml.v2",
        "organization": "gopkg.in",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "2.2.1",
      },
      Object {
        "module": "api",
        "organization": "istio.io",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "api",
        "organization": "k8s.io",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "release-1.11",
      },
      Object {
        "module": "apimachinery",
        "organization": "k8s.io",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "release-1.11",
      },
      Object {
        "module": "apiextensions-apiserver",
        "organization": "k8s.io",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "release-1.11",
      },
      Object {
        "module": "helm",
        "organization": "k8s.io",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "2.9.1",
      },
      Object {
        "module": "kubernetes",
        "organization": "k8s.io",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "release-1.11",
      },
      Object {
        "module": "gogo/status",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "d60b5acac426cb52f7e28124c6aaa862d9f339f5",
      },
      Object {
        "module": "google/go-cmp",
        "organization": "github.com",
        "scopes": Array [
          "constraint",
        ],
        "versionConstraint": "0.2.0",
      },
      Object {
        "module": "docker/distribution",
        "organization": "github.com",
        "scopes": Array [
          "override",
        ],
        "versionConstraint": "master",
      },
      Object {
        "module": "client-go",
        "organization": "k8s.io",
        "scopes": Array [
          "override",
        ],
        "versionConstraint": "release-8.0",
      },
      Object {
        "module": "Azure/go-autorest",
        "organization": "github.com",
        "scopes": Array [
          "override",
        ],
        "versionConstraint": "bca49d5b51a50dc5bb17bbf6204c711c6dbded06",
      },
      Object {
        "module": "golang/glog",
        "organization": "github.com",
        "scopes": Array [
          "override",
        ],
        "versionConstraint": undefined,
      },
      Object {
        "module": "imdario/mergo",
        "organization": "github.com",
        "scopes": Array [
          "override",
        ],
        "versionConstraint": "6633656539c1639d9d78127b7d47c622b5d7b6dc",
      },
      Object {
        "module": "istio/mixer/adapter/stackdriver/contextgraph",
        "organization": "istio.io",
        "scopes": Array [
          "ignored",
        ],
        "versionConstraint": "*",
      },
      Object {
        "module": "client9/misspell",
        "organization": "github.com",
        "scopes": Array [
          "required",
        ],
        "versionConstraint": "*",
      },
      Object {
        "module": "istio/tools/protoc-gen-docs",
        "organization": "github.com",
        "scopes": Array [
          "required",
        ],
        "versionConstraint": "*",
      },
      Object {
        "module": "fortio",
        "organization": "fortio.org",
        "scopes": Array [
          "required",
        ],
        "versionConstraint": "*",
      },
    ],
    "language": "go",
    "module": "deps-cloud/extractor",
    "organization": "github.com",
    "system": "gopkg",
    "version": "latest",
  },
  Object {
    "dependencies": Array [
      Object {
        "module": "jcommander",
        "organization": "com.beust",
        "scopes": Array [
          "default",
        ],
        "versionConstraint": "1.20",
      },
      Object {
        "module": "lombok",
        "organization": "org.projectlombok",
        "scopes": Array [
          "build->default",
        ],
        "versionConstraint": "0.10.4",
      },
      Object {
        "module": "junit",
        "organization": "junit",
        "scopes": Array [
          "test->default",
        ],
        "versionConstraint": "4.10",
      },
      Object {
        "module": "cobertura",
        "organization": "net.sourceforge.cobertura",
        "scopes": Array [
          "test->default",
        ],
        "versionConstraint": "1.9.4.1",
      },
      Object {
        "module": "checkstyle",
        "organization": "com.puppycrawl.tools",
        "scopes": Array [
          "analysis->default",
        ],
        "versionConstraint": "5.5",
      },
      Object {
        "module": "findbugs",
        "organization": "com.google.code.findbugs",
        "scopes": Array [
          "analysis->default",
        ],
        "versionConstraint": "2.0.0",
      },
      Object {
        "module": "pmd",
        "organization": "net.sourceforge.pmd",
        "scopes": Array [
          "analysis->default",
        ],
        "versionConstraint": "5.0.0",
      },
    ],
    "language": "java",
    "module": "sample-java-project",
    "organization": "com.nullprogram",
    "system": "ivy",
    "version": null,
  },
  Object {
    "dependencies": Array [
      Object {
        "module": "runtimeDependency",
        "organization": "scoped",
        "scopes": Array [
          "",
        ],
        "versionConstraint": "1.0.0",
      },
      Object {
        "module": "devDependency",
        "organization": "_",
        "scopes": Array [
          "dev",
        ],
        "versionConstraint": "1.0.0",
      },
      Object {
        "module": "peerDependency",
        "organization": "scoped",
        "scopes": Array [
          "peer",
        ],
        "versionConstraint": "1.0.0",
      },
      Object {
        "module": "optionalDependency",
        "organization": "_",
        "scopes": Array [
          "optional",
        ],
        "versionConstraint": "1.0.0",
      },
    ],
    "language": "node",
    "module": "module",
    "organization": "organization",
    "system": "npm",
    "version": "9.9.9",
  },
  Object {
    "dependencies": Array [
      Object {
        "module": "parentArtifactId",
        "organization": "parentGroupId",
        "scopes": Array [
          "parent",
        ],
        "versionConstraint": "1.0",
      },
      Object {
        "module": "compileArtifactId",
        "organization": "compileGroupId",
        "scopes": Array [
          "compile",
        ],
        "versionConstraint": "1.0.0",
      },
      Object {
        "module": "runtimeArtifactId",
        "organization": "runtimeGroupId",
        "scopes": Array [
          "runtime",
        ],
        "versionConstraint": "1.0.0",
      },
      Object {
        "module": "testArtifactId",
        "organization": "testGroupId",
        "scopes": Array [
          "test",
        ],
        "versionConstraint": "1.0.0",
      },
      Object {
        "module": "systemArtifactId",
        "organization": "systemGroupId",
        "scopes": Array [
          "system",
        ],
        "versionConstraint": "1.0.0",
      },
      Object {
        "module": "providedArtifactId",
        "organization": "providedGroupId",
        "scopes": Array [
          "provided",
        ],
        "versionConstraint": "1.0.0",
      },
    ],
    "language": "java",
    "module": "artifactId",
    "organization": "groupId",
    "system": "maven",
    "version": "9.9.9",
  },
]
`;
